File: ios/LookSeeProto/UI Files/BusinessMediaHistoryView.swift
//
//  BusinessMediaHistoryView.swift
//  LookSeeProto
//
//  Displays upload history for one business-owned landmark.
//

import SwiftUI

struct BusinessMediaHistoryView: View {
    @StateObject private var viewModel: BusinessMediaHistoryViewModel

    init(
        landmarkId: String,
        landmarkLabel: String
    ) {
        _viewModel = StateObject(
            wrappedValue: BusinessMediaHistoryViewModel(
                landmarkId: landmarkId,
                landmarkLabel: landmarkLabel
            )
        )
    }

    var body: some View {
        Group {
            if viewModel.isLoadingInitial && viewModel.items.isEmpty {
                initialLoadingView
            } else if viewModel.items.isEmpty {
                emptyOrErrorView
            } else {
                historyList
            }
        }
        .navigationTitle("Media History")
        .navigationBarTitleDisplayMode(.inline)
        .task {
            await viewModel.loadInitial()
        }
        .task(id: viewModel.processingPollKey) {
            await viewModel.pollProcessingItems()
        }
        .toolbar {
            ToolbarItem(placement: .navigationBarTrailing) {
                Button {
                    Task {
                        await viewModel.refresh()
                    }
                } label: {
                    if viewModel.isRefreshing {
                        ProgressView()
                    } else {
                        Image(systemName: "arrow.clockwise")
                    }
                }
                .disabled(
                    viewModel.isRefreshing ||
                    viewModel.isLoadingInitial ||
                    viewModel.isLoadingMore
                )
                .accessibilityLabel("Refresh media history")
            }
        }
    }

    private var historyList: some View {
        List {
            Section {
                VStack(alignment: .leading, spacing: 5) {
                    Text(viewModel.landmarkLabel)
                        .font(.headline)

                    Text("\(viewModel.items.count) upload\(viewModel.items.count == 1 ? "" : "s") loaded")
                        .font(.caption)
                        .foregroundColor(.secondary)
                }
                .padding(.vertical, 2)
            }

            Section(header: Text("Uploads")) {
                ForEach(viewModel.items) { item in
                    BusinessMediaHistoryRow(
                        item: item,
                        isRetrying: viewModel.isRetrying(item),
                        retryError: viewModel.retryError(for: item)
                    ) {
                        Task {
                            await viewModel.retryProcessing(item)
                        }
                    }
                }

                if viewModel.hasMoreItems {
                    Button {
                        Task {
                            await viewModel.loadMore()
                        }
                    } label: {
                        HStack {
                            Spacer()

                            if viewModel.isLoadingMore {
                                ProgressView()
                                Text("Loading more...")
                            } else {
                                Label(
                                    "Load More",
                                    systemImage: "arrow.down.circle"
                                )
                            }

                            Spacer()
                        }
                    }
                    .disabled(viewModel.isLoadingMore)
                }
            }

            if let errorMessage = viewModel.errorMessage {
                Section {
                    errorMessageView(errorMessage)
                }
            }
        }
        .listStyle(.insetGrouped)
        .refreshable {
            await viewModel.refresh()
        }
    }

    private var initialLoadingView: some View {
        VStack(spacing: 12) {
            ProgressView()
            Text("Loading media history...")
                .font(.subheadline)
                .foregroundColor(.secondary)
        }
        .frame(maxWidth: .infinity, maxHeight: .infinity)
    }

    private var emptyOrErrorView: some View {
        VStack(spacing: 16) {
            Image(
                systemName: viewModel.errorMessage == nil
                    ? "photo.on.rectangle.angled"
                    : "exclamationmark.triangle"
            )
            .font(.system(size: 42))
            .foregroundColor(.secondary)

            Text(
                viewModel.errorMessage == nil
                    ? "No Upload History"
                    : "Couldn’t Load History"
            )
            .font(.title3.weight(.semibold))

            Text(
                viewModel.errorMessage
                    ?? "New positive and negative uploads for this landmark will appear here."
            )
            .font(.subheadline)
            .foregroundColor(.secondary)
            .multilineTextAlignment(.center)
            .padding(.horizontal, 28)

            if viewModel.errorMessage != nil {
                Button("Try Again") {
                    Task {
                        await viewModel.retry()
                    }
                }
                .buttonStyle(.borderedProminent)
            }
        }
        .frame(maxWidth: .infinity, maxHeight: .infinity)
    }

    private func errorMessageView(_ message: String) -> some View {
        VStack(alignment: .leading, spacing: 10) {
            Label(
                "Some history could not be loaded",
                systemImage: "exclamationmark.triangle.fill"
            )
            .foregroundColor(.orange)

            Text(message)
                .font(.footnote)
                .foregroundColor(.secondary)

            Button("Retry") {
                Task {
                    await viewModel.retry()
                }
            }
        }
        .padding(.vertical, 4)
    }
}

private struct BusinessMediaHistoryRow: View {
    let item: BusinessMediaHistoryItem
    let isRetrying: Bool
    let retryError: String?
    let onRetry: () -> Void

    var body: some View {
        HStack(alignment: .top, spacing: 14) {
            thumbnail

            VStack(alignment: .leading, spacing: 7) {
                Text(item.roleAndMediaTitle)
                    .font(.headline)
                    .lineLimit(2)

                statusBadge

                Label(
                    item.uploadedBy.displayText,
                    systemImage: "person.crop.circle"
                )
                .font(.subheadline)
                .foregroundColor(.secondary)
                .lineLimit(1)

                Label(uploadDateText, systemImage: "calendar")
                .font(.subheadline)
                .foregroundColor(.secondary)

                VStack(alignment: .leading, spacing: 2) {
                    Text("Submission ID")
                        .font(.caption2)
                        .foregroundColor(.secondary)

                    Text(item.submissionId)
                        .font(.caption.monospaced())
                        .foregroundColor(.primary)
                        .lineLimit(1)
                        .truncationMode(.middle)
                        .textSelection(.enabled)
                }

                Text(item.displayFilename)
                    .font(.caption)
                    .foregroundColor(.secondary)
                    .lineLimit(1)
                    .truncationMode(.middle)

                if item.isProcessingDelayed {
                    Label(
                        "Processing longer than expected",
                        systemImage: "exclamationmark.circle.fill"
                    )
                    .font(.caption2.weight(.semibold))
                    .foregroundColor(.orange)
                }

                if item.canRetryProcessing {
                    Button(action: onRetry) {
                        HStack(spacing: 7) {
                            if isRetrying {
                                ProgressView()
                                    .controlSize(.small)
                            } else {
                                Image(systemName: "arrow.clockwise.circle.fill")
                            }
                            Text(isRetrying ? "Requeueing..." : "Retry Processing")
                        }
                        .font(.caption.weight(.semibold))
                    }
                    .buttonStyle(.bordered)
                    .disabled(isRetrying)
                }

                if let retryCount = item.retryCount, retryCount > 0 {
                    Text("Processing retried \(retryCount) time\(retryCount == 1 ? "" : "s")")
                        .font(.caption2)
                        .foregroundColor(.secondary)
                }

                if let retryError {
                    Text(retryError)
                        .font(.caption2)
                        .foregroundColor(.red)
                }
            }
        }
        .padding(.vertical, 7)
    }

    private var thumbnail: some View {
        Group {
            if let thumbnailUrl = item.thumbnailUrl {
                AsyncImage(url: thumbnailUrl) { phase in
                    switch phase {
                    case .empty:
                        thumbnailPlaceholder(showProgress: true)
                    case .success(let image):
                        image
                            .resizable()
                            .scaledToFill()
                    case .failure:
                        thumbnailPlaceholder(showProgress: false)
                    @unknown default:
                        thumbnailPlaceholder(showProgress: false)
                    }
                }
            } else {
                thumbnailPlaceholder(showProgress: false)
            }
        }
        .frame(width: 88, height: 72)
        .background(Color.secondary.opacity(0.12))
        .clipShape(RoundedRectangle(cornerRadius: 10))
        .overlay(alignment: .bottomTrailing) {
            if item.isVideo {
                Image(systemName: "video.fill")
                    .font(.caption2)
                    .padding(5)
                    .foregroundColor(.white)
                    .background(Color.black.opacity(0.65))
                    .clipShape(Circle())
                    .padding(5)
            }
        }
    }

    private func thumbnailPlaceholder(showProgress: Bool) -> some View {
        ZStack {
            Color.secondary.opacity(0.10)

            if showProgress {
                ProgressView()
            } else {
                Image(systemName: item.mediaSystemImage)
                    .font(.title2)
                    .foregroundColor(.secondary)
            }
        }
    }

    private var statusBadge: some View {
        Label(item.displayStatus, systemImage: statusSystemImage)
            .font(.caption2.weight(.semibold))
            .padding(.horizontal, 7)
            .padding(.vertical, 4)
            .foregroundColor(statusColor)
            .background(statusColor.opacity(0.14))
            .clipShape(Capsule())
            .accessibilityLabel(
                "Media status: \(item.displayStatus). Backend status: \(item.backendStatusText)"
            )
    }

    private var statusColor: Color {
        switch item.lifecycleState {
        case .ready: return .green
        case .processing: return .orange
        case .failed: return .red
        case .unknown: return .secondary
        }
    }

    private var statusSystemImage: String {
        if item.isProcessingDelayed {
            return "exclamationmark.circle.fill"
        }

        switch item.lifecycleState {
        case .ready: return "checkmark.circle.fill"
        case .processing: return "clock.arrow.circlepath"
        case .failed: return "exclamationmark.triangle.fill"
        case .unknown: return "questionmark.circle"
        }
    }

    private var uploadDateText: String {
        guard let uploadDate = item.uploadDate else {
            return "Date unavailable"
        }
        return uploadDate.formatted(date: .abbreviated, time: .shortened)
    }
}

File: Android/LookSee/app/src/main/java/looksee/angelll/com/uifiles/BusinessMediaHistoryView.kt
package looksee.angelll.com.uifiles

import androidx.compose.foundation.background
import androidx.compose.foundation.layout.*
import androidx.compose.foundation.lazy.LazyColumn
import androidx.compose.foundation.lazy.items
import androidx.compose.foundation.shape.CircleShape
import androidx.compose.foundation.shape.RoundedCornerShape
import androidx.compose.material.icons.Icons
import androidx.compose.material.icons.filled.*
import androidx.compose.material3.*
import androidx.compose.runtime.*
import androidx.compose.ui.Alignment
import androidx.compose.ui.Modifier
import androidx.compose.ui.draw.clip
import androidx.compose.ui.graphics.Color
import androidx.compose.ui.text.font.FontWeight
import androidx.compose.ui.unit.dp
import androidx.compose.ui.unit.sp
import kotlinx.coroutines.launch
import looksee.angelll.com.models.*
import looksee.angelll.com.viewmodels.*
import looksee.angelll.com.services.*
import coil.compose.AsyncImage
import java.time.ZoneId
import java.time.format.DateTimeFormatter
import java.time.format.FormatStyle

@OptIn(ExperimentalMaterial3Api::class)
@Composable
fun BusinessMediaHistoryView(
    landmarkId: String,
    landmarkLabel: String
) {
    val viewModel = remember { BusinessMediaHistoryViewModel(landmarkId, landmarkLabel) }
    val items by viewModel.items.collectAsState()
    val isLoadingInitial by viewModel.isLoadingInitial.collectAsState()
    val isRefreshing by viewModel.isRefreshing.collectAsState()
    val isLoadingMore by viewModel.isLoadingMore.collectAsState()
    val errorMessage by viewModel.errorMessage.collectAsState()
    val hasMoreItems = viewModel.hasMoreItems
    val landmarkTitle by viewModel.landmarkLabel.collectAsState()

    val coroutineScope = rememberCoroutineScope()

    LaunchedEffect(Unit) {
        viewModel.loadInitial()
    }

    LaunchedEffect(viewModel.processingPollKey) {
        viewModel.pollProcessingItems()
    }

    Scaffold(
        topBar = {
            TopAppBar(
                title = { Text("Upload History") },
                actions = {
                    IconButton(onClick = { coroutineScope.launch { viewModel.refresh() } }, enabled = !isRefreshing) {
                        if (isRefreshing) CircularProgressIndicator(modifier = Modifier.size(24.dp), strokeWidth = 2.dp)
                        else Icon(Icons.Default.Refresh, contentDescription = "Refresh")
                    }
                }
            )
        },
        containerColor = Color(0xFFF2F2F7)
    ) { paddingValues ->
        if (isLoadingInitial) {
            Box(modifier = Modifier.fillMaxSize().padding(paddingValues), contentAlignment = Alignment.Center) {
                CircularProgressIndicator()
            }
        } else if (errorMessage != null && items.isEmpty()) {
            Box(modifier = Modifier.fillMaxSize().padding(paddingValues), contentAlignment = Alignment.Center) {
                Column(horizontalAlignment = Alignment.CenterHorizontally) {
                    Text(errorMessage!!, color = Color.Red, modifier = Modifier.padding(16.dp))
                    Button(onClick = { coroutineScope.launch { viewModel.retry() } }) {
                        Text("Retry")
                    }
                }
            }
        } else {
            LazyColumn(
                modifier = Modifier.fillMaxSize().padding(paddingValues).padding(horizontal = 16.dp),
                verticalArrangement = Arrangement.spacedBy(16.dp)
            ) {
                // Header Section
                item {
                    Column(verticalArrangement = Arrangement.spacedBy(5.dp), modifier = Modifier.padding(vertical = 2.dp)) {
                        Text(landmarkTitle, fontWeight = FontWeight.Bold, fontSize = 18.sp)
                        Text(
                            text = "${items.size} upload${if (items.size == 1) "" else "s"} loaded",
                            fontSize = 12.sp,
                            color = Color.Gray
                        )
                    }
                }

                // Uploads Section
                item {
                    Text("UPLOADS", fontSize = 13.sp, fontWeight = FontWeight.Bold, color = Color.Gray)
                }

                items(items) { item ->
                    Surface(
                        shape = RoundedCornerShape(12.dp),
                        color = Color.White,
                        modifier = Modifier.fillMaxWidth(),
                        shadowElevation = 1.dp
                    ) {
                        BusinessMediaHistoryRow(
                            item = item,
                            isRetrying = viewModel.isRetrying(item),
                            retryError = viewModel.retryError(item),
                            onRetry = { coroutineScope.launch { viewModel.retryProcessing(item) } }
                        )
                    }
                }

                // Load More Section
                if (hasMoreItems) {
                    item {
                        TextButton(
                            onClick = { coroutineScope.launch { viewModel.loadMore() } },
                            enabled = !isLoadingMore,
                            modifier = Modifier.fillMaxWidth()
                        ) {
                            if (isLoadingMore) CircularProgressIndicator(modifier = Modifier.size(16.dp), strokeWidth = 2.dp)
                            else Text("Load More", color = Color(0xFF007AFF))
                        }
                    }
                }

                if (errorMessage != null && items.isNotEmpty()) {
                    item {
                        Text(errorMessage!!, color = Color.Red, fontSize = 13.sp, modifier = Modifier.padding(8.dp))
                        Button(onClick = { coroutineScope.launch { viewModel.retry() } }, modifier = Modifier.fillMaxWidth()) {
                            Text("Retry Loading More")
                        }
                    }
                }
            }
        }
    }
}

@Composable
fun BusinessMediaHistoryRow(
    item: BusinessMediaHistoryItem,
    isRetrying: Boolean,
    retryError: String?,
    onRetry: () -> Unit
) {
    Row(
        modifier = Modifier
            .padding(14.dp)
            .fillMaxWidth(),
        horizontalArrangement = Arrangement.spacedBy(14.dp),
        verticalAlignment = Alignment.Top
    ) {
        BusinessMediaThumbnail(item)

        Column(modifier = Modifier.weight(1f), verticalArrangement = Arrangement.spacedBy(7.dp)) {
            Text(
                text = item.roleAndMediaTitle,
                fontWeight = FontWeight.Bold,
                fontSize = 17.sp,
                lineHeight = 22.sp
            )

            BusinessMediaStatusBadge(item)

            Row(verticalAlignment = Alignment.CenterVertically, horizontalArrangement = Arrangement.spacedBy(6.dp)) {
                Icon(Icons.Default.AccountCircle, contentDescription = null, tint = Color.Gray, modifier = Modifier.size(14.dp))
                Text(item.uploadedBy.displayText, fontSize = 13.sp, color = Color.Gray)
            }

            Row(verticalAlignment = Alignment.CenterVertically, horizontalArrangement = Arrangement.spacedBy(6.dp)) {
                Icon(Icons.Default.CalendarToday, contentDescription = null, tint = Color.Gray, modifier = Modifier.size(14.dp))
                val dateText = remember(item.uploadInstant) {
                    item.uploadInstant?.let {
                        DateTimeFormatter.ofLocalizedDateTime(FormatStyle.MEDIUM)
                            .withZone(ZoneId.systemDefault())
                            .format(it)
                    } ?: item.uploadedAtISO ?: "Date unavailable"
                }
                Text(dateText, fontSize = 13.sp, color = Color.Gray)
            }

            Column(verticalArrangement = Arrangement.spacedBy(2.dp)) {
                Text("Submission ID", fontSize = 11.sp, color = Color.Gray)
                Text(
                    text = item.submissionId,
                    fontSize = 12.sp,
                    fontFamily = androidx.compose.ui.text.font.FontFamily.Monospace,
                    color = Color.White,
                    maxLines = 1
                )
            }

            if (item.displayFilename.isNotEmpty()) {
                Text(
                    text = item.displayFilename,
                    fontSize = 12.sp,
                    color = Color.Gray,
                    maxLines = 1
                )
            }

            if (item.isProcessingDelayed) {
                Row(
                    verticalAlignment = Alignment.CenterVertically,
                    horizontalArrangement = Arrangement.spacedBy(6.dp)
                ) {
                    Icon(Icons.Default.Error, contentDescription = null, tint = Color(0xFFFFA500), modifier = Modifier.size(14.dp))
                    Text(
                        "Processing longer than expected",
                        fontSize = 11.sp,
                        fontWeight = FontWeight.SemiBold,
                        color = Color(0xFFFFA500)
                    )
                }
            }

            if (item.canRetryProcessing) {
                Button(
                    onClick = onRetry,
                    enabled = !isRetrying,
                    colors = ButtonDefaults.buttonColors(
                        containerColor = Color.White.copy(alpha = 0.1f),
                        contentColor = Color.White
                    ),
                    shape = RoundedCornerShape(8.dp),
                    contentPadding = PaddingValues(horizontal = 12.dp, vertical = 6.dp),
                    modifier = Modifier.height(32.dp)
                ) {
                    Row(verticalAlignment = Alignment.CenterVertically, horizontalArrangement = Arrangement.spacedBy(6.dp)) {
                        if (isRetrying) {
                            CircularProgressIndicator(modifier = Modifier.size(12.dp), strokeWidth = 2.dp, color = Color.White)
                        } else {
                            Icon(Icons.Default.Refresh, contentDescription = null, modifier = Modifier.size(14.dp))
                        }
                        Text(if (isRetrying) "Requeueing..." else "Retry Processing", fontSize = 12.sp, fontWeight = FontWeight.SemiBold)
                    }
                }
            }

            if (item.retryCount != null && item.retryCount > 0) {
                Text(
                    "Processing retried ${item.retryCount} time${if (item.retryCount == 1) "" else "s"}",
                    fontSize = 11.sp,
                    color = Color.Gray
                )
            }

            if (retryError != null) {
                Text(retryError, fontSize = 11.sp, color = Color.Red)
            }
        }
    }
}

@Composable
fun BusinessMediaThumbnail(item: BusinessMediaHistoryItem) {
    Box(
        modifier = Modifier
            .size(width = 88.dp, height = 72.dp)
            .clip(RoundedCornerShape(10.dp))
            .background(Color.White.copy(alpha = 0.12f)),
        contentAlignment = Alignment.Center
    ) {
        val thumb = item.thumbnailUrl
        if (!thumb.isNullOrEmpty()) {
            AsyncImage(
                model = thumb,
                contentDescription = null,
                modifier = Modifier.fillMaxSize(),
                contentScale = androidx.compose.ui.layout.ContentScale.Crop
            )
        }
 else {
            Icon(
                imageVector = if (item.isVideo) Icons.Default.Videocam else Icons.Default.Image,
                contentDescription = null,
                tint = Color.Gray,
                modifier = Modifier.size(24.dp)
            )
        }

        if (item.isVideo) {
            Box(
                modifier = Modifier
                    .align(Alignment.BottomEnd)
                    .padding(5.dp)
                    .size(20.dp)
                    .background(Color.Black.copy(alpha = 0.65f), CircleShape),
                contentAlignment = Alignment.Center
            ) {
                Icon(Icons.Default.PlayArrow, contentDescription = null, tint = Color.White, modifier = Modifier.size(12.dp))
            }
        }
    }
}


@Composable
fun BusinessMediaStatusBadge(item: BusinessMediaHistoryItem) {
    val statusColor = when (item.lifecycleState) {
        BusinessMediaLifecycleState.READY -> Color(0xFF34C759)
        BusinessMediaLifecycleState.PROCESSING -> Color(0xFF007AFF)
        BusinessMediaLifecycleState.FAILED -> Color.Red
        BusinessMediaLifecycleState.UNKNOWN -> Color.Gray
    }

    val statusIcon = when (item.lifecycleState) {
        BusinessMediaLifecycleState.READY -> Icons.Default.CheckCircle
        BusinessMediaLifecycleState.PROCESSING -> Icons.Default.HourglassEmpty
        BusinessMediaLifecycleState.FAILED -> Icons.Default.Error
        BusinessMediaLifecycleState.UNKNOWN -> Icons.Default.QuestionMark
    }

    Row(
        modifier = Modifier.background(statusColor.copy(alpha = 0.1f), RoundedCornerShape(12.dp)).padding(horizontal = 8.dp, vertical = 4.dp),
        verticalAlignment = Alignment.CenterVertically,
        horizontalArrangement = Arrangement.spacedBy(4.dp)
    ) {
        Icon(statusIcon, contentDescription = null, tint = statusColor, modifier = Modifier.size(12.dp))
        Text(item.displayStatus, color = statusColor, fontWeight = FontWeight.Bold, fontSize = 11.sp)
    }
}
